rust/gitxetcore/src/xetblob/file_operations/cp_operations.rs

```diff
@@ -0,0 +1,469 @@
+                } else if dest_specified_as_directory {
+                    // Slightly different behavior depending on whether the destination is specified as
+                    // a directory or not.  If it is, then copy the source file into the dest path, otherwise
+                    // copy the source to the dest path name.
+
+                    // git-xet cp dir/subdir/file dest/d1/  -> goes into dest/d1/file
+
+                    dest_dir = dest_base_path.to_owned();
+                    let end_component = src_fs.file_name(&src_path);
+                    dest_path = dest_fs.join(&dest_dir, end_component);
+                } else {
+                    // git-xet cp dir/subdir/file dest/bar  -> goes into dest/bar
+                    dest_dir = dest_fs.parent(dest_base_path).unwrap_or("").to_owned();
+                    dest_path = dest_base_path.to_owned();
+                }
+            }
+        };
+
+        // Now, we should have all the variables -- src_is_directory, src_path, dest_path, dest_dir, and file_filter_pattern
+        // set up correctly.  The easiest way to break this up is between the multiple file case (src_is_directory = True) and the
+        // single file case.
+
+        if src_is_directory {
+            // With the source a directory, we need to list out all the files to copy.
+            if recursive {
+                // If both the source and the dest are xet, then
+                if src_fs.is_xet() && dest_fs.is_xet() {
+                    // In this case, this can be just one operation.
+                    cp_ops.push(CPOperation {
+                        src_path,
+                        dest_dir,
+                        dest_path,
+                        size: 0,
+                    });
+                } else {
+                    cp_ops.extend(src_fs.listdir(&src_path, true).await?.into_iter().map(
+                        |entry| {
+                            let cp_src_path = src_fs.join(&src_path, &entry.name);
+                            let cp_dest_path = dest_fs.join(&dest_path, &entry.name).to_owned();
+                            let cp_dest_dir = dest_fs
+                                .parent(&cp_dest_path)
+                                .unwrap_or(&dest_dir)
+                                .to_owned();
+
+                            CPOperation {
+                                src_path: cp_src_path,
+                                dest_path: cp_dest_path,
+                                dest_dir: cp_dest_dir,
+                                size: entry.size,
+                            }
+                        },
+                    ));
+                }
+            } else {
+                return Err(GitXetRepoError::InvalidOperation(format!(
+                    "Copy: source {src} is a directory, but cp not recursive."
+                )));
+            }
+        } else {
+            cp_ops.push(CPOperation {
```

Comment:
If src is file and dest is directory do you need to take the src filename and append to dest?  (I am still not sure what is dest_dir and dest_path)

Reply:
That is done above.